test/Constraints/array_literal.swift

```diff
@@ -145,16 +145,48 @@ func defaultToAny(i: Int, s: String) {
   let _: [Any] = [1, "a", 3.5]
   let _: [Any] = [1, "a", [3.5, 3.7, 3.9]]
   let _: [Any] = [1, "a", [3.5, "b", 3]]
+  // expected-warning@-1{{heterogeneous collection literal could only be inferred to '[Any]'; add explicit type annotation if this is intentional}}
   let _: [Any] = [1, [2, [3]]]
+  // expected-warning@-1{{heterogeneous collection literal could only be inferred to '[Any]'; add explicit type annotation if this is intentional}}
   
   let _: [Any?] = [1, "a", nil, 3.5]
   let _: [Any?] = [1, "a", nil, [3.5, 3.7, 3.9]]
   let _: [Any?] = [1, "a", nil, [3.5, "b", nil]]
+  // expected-warning@-1{{heterogeneous collection literal could only be inferred to '[Any?]'; add explicit type annotation if this is intentional}}
   let _: [Any?] = [1, [2, [3]]]
+  // expected-warning@-1{{heterogeneous collection literal could only be inferred to '[Any]'; add explicit type annotation if this is intentional}}
```

Comment:
@simanerush @AnthonyLatsis I think we don't want to diagnose cases like this one (note that it's different from the following one where inner collection literal is heterogeneous) and the ones where `[]` is used in `return` i.e. https://github.com/apple/swift/pull/60323/files#diff-55b6c1da2887f7323cf5ddb64f0a2add1f0b6a1392a51113d20bbbd04f8ccfc7R12 because there is a contextual type in this case so it's equivalent to `as [Any]` or `as [Any?]` which we suggest to silence the warning.

Reply:
I think the compiler is correct in this particular case because there is no explicit element type specified for `[2, [3]]`, which is defaulted to `[Any]` and coerced to `Any?` as an element of the outer `[Any?]` literal.

Reply:
Agree. This change originated from @xedin's point earlier on about nested collection literals with a defaulted type.

Reply:
Ah yes, I missed that, sorry, this one is indeed correct

Reply:
IMO we are also correct to warn on `return []` for the same reason. `func f() -> some Collection { return [] }` is analogous to `let _: some Collection = []`.

Reply:
Would we warn about `let _: [Any] = []`?

Reply:
`some Collection` case is different really because the underlying type of opaque has to be inferred from the result expression, so it's reasonable to warn there but in case where result is a concrete type I don't think so.

Reply:
@xedin I can't find an example of emission of warning diagnostics in a case of a function with a concrete type -- all tests have an opaque type. Could you maybe point to a case where we emit diagnostics with a concrete type function?

Reply:
> Would we warn about `let _: [Any] = []`?

We should not.

Reply:
@simanerush It must have been an older commit or something but I could swear I saw a test-case that got updated with the warning that returned `[Any]` similar to this one:

```
func test() -> [Any] {
  []
}
```

Reply:
Since it's not in the changes anymore I think we can land this with a `// FIXME` comment for all the test-cases mentioned in https://github.com/apple/swift/pull/60323/files#r942866744. That could be addressed separately.

Reply:
> ```swift
> func test() -> [Any] {
>   []
> }
> ```
@simanerush Could you add test cases for this and `let _: [Any] = []` as you fix the conflict?

Reply:
@AnthonyLatsis sure!